### PR TITLE
Calculate MD5 checksum of MessageAttributes

### LINKED ISSUE #104: MD5 checksum of message attributes incorrect
It seems that the checksum generation of SQS message attributes is not implemented, correct? In
https://github.com/p4tin/goaws/blob/c4d80809ebe7f664226bd62fc60f355ae7a8b8ed/app/gosqs/gosqs.go#L131 the checksum of the message attributes is always computed from the static string `GoAws`. 

When using the java sdk when talking to SQS, sending a message to the queue fails because the client computes a different checksum for the message attributes than what goaws returns.

```
    private static void sendMessageOperationMd5Check(SendMessageRequest sendMessageRequest, SendMessageResult sendMessageResult) {
        String messageBodySent = sendMessageRequest.getMessageBody();
        String bodyMd5Returned = sendMessageResult.getMD5OfMessageBody();
        String clientSideBodyMd5 = calculateMessageBodyMd5(messageBodySent);
        if(!clientSideBodyMd5.equals(bodyMd5Returned)) {
            throw new AmazonClientException(String.format("MD5 returned by SQS does not match the calculation on the original request. (MD5 calculated by the %s: \"%s\", MD5 checksum returned: \"%s\")", new Object[]{"message body", clientSideBodyMd5, bodyMd5Returned}));
        } else {
            Map<String, MessageAttributeValue> messageAttrSent = sendMessageRequest.getMessageAttributes();
            if(messageAttrSent != null && !messageAttrSent.isEmpty()) {
                String clientSideAttrMd5 = calculateMessageAttributesMd5(messageAttrSent);
                String attrMd5Returned = sendMessageResult.getMD5OfMessageAttributes();
                if(!clientSideAttrMd5.equals(attrMd5Returned)) {
                    throw new AmazonClientException(String.format("MD5 returned by SQS does not match the calculation on the original request. (MD5 calculated by the %s: \"%s\", MD5 checksum returned: \"%s\")", new Object[]{"message attributes", clientSideAttrMd5, attrMd5Returned}));
                }
            }

        }
    }
```

```
com.amazonaws.AmazonClientException: MD5 returned by SQS does not match the calculation on the original request. (MD5 calculated by the message attributes: "3226c32921e2bcda84eed17654164acc", MD5 checksum returned: "b095c6d16871105acb75d59332513337")
	at com.amazonaws.services.sqs.MessageMD5ChecksumHandler.sendMessageOperationMd5Check(MessageMD5ChecksumHandler.java:120) ~[aws-java-sdk-sqs-1.11.98.jar:na]
	at com.amazonaws.services.sqs.MessageMD5ChecksumHandler.afterResponse(MessageMD5ChecksumHandler.java:80) ~[aws-java-sdk-sqs-1.11.98.jar:na]
	at com.amazonaws.handlers.RequestHandler2Adaptor.afterResponse(RequestHandler2Adaptor.java:49) ~[aws-java-sdk-core-1.11.98.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.afterResponse(AmazonHttpClient.java:971) ~[aws-java-sdk-core-1.11.98.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:745) ~[aws-java-sdk-core-1.11.98.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:716) ~[aws-java-sdk-core-1.11.98.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699) ~[aws-java-sdk-core-1.11.98.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667) ~[aws-java-sdk-core-1.11.98.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649) ~[aws-java-sdk-core-1.11.98.jar:na]
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513) ~[aws-java-sdk-core-1.11.98.jar:na]
	at com.amazonaws.services.sqs.AmazonSQSClient.doInvoke(AmazonSQSClient.java:1689) ~[aws-java-sdk-sqs-1.11.98.jar:na]
	at com.amazonaws.services.sqs.AmazonSQSClient.invoke(AmazonSQSClient.java:1665) ~[aws-java-sdk-sqs-1.11.98.jar:na]
	at com.amazonaws.services.sqs.AmazonSQSClient.sendMessage(AmazonSQSClient.java:1435) ~[aws-java-sdk-sqs-1.11.98.jar:na]
	at com.amazon.sqs.javamessaging.AmazonSQSMessagingClientWrapper.sendMessage(AmazonSQSMessagingClientWrapper.java:168) ~[amazon-sqs-java-messaging-lib-1.0.2.jar:na]
	at com.amazon.sqs.javamessaging.SQSMessageProducer.sendInternal(SQSMessageProducer.java:111) ~[amazon-sqs-java-messaging-lib-1.0.2.jar:na]
	at com.amazon.sqs.javamessaging.SQSMessageProducer.send(SQSMessageProducer.java:291) ~[amazon-sqs-java-messaging-lib-1.0.2.jar:na]
	at org.springframework.jms.core.JmsTemplate.doSend(JmsTemplate.java:626) ~[spring-jms-4.3.7.RELEASE.jar:4.3.7.RELEASE]
	at org.springframework.jms.core.JmsTemplate.doSend(JmsTemplate.java:597) ~[spring-jms-4.3.7.RELEASE.jar:4.3.7.RELEASE]
	at org.springframework.jms.core.JmsTemplate$4.doInJms(JmsTemplate.java:574) ~[spring-jms-4.3.7.RELEASE.jar:4.3.7.RELEASE]
	at org.springframework.jms.core.JmsTemplate.execute(JmsTemplate.java:484) ~[spring-jms-4.3.7.RELEASE.jar:4.3.7.RELEASE]
	at org.springframework.jms.core.JmsTemplate.send(JmsTemplate.java:570) ~[spring-jms-4.3.7.RELEASE.jar:4.3.7.RELEASE]
	at org.springframework.jms.core.JmsTemplate.convertAndSend(JmsTemplate.java:658) ~[spring-jms-4.3.7.RELEASE.jar:4.3.7.RELEASE]
```

### PR DESCRIPTION
**Pull request by ulich**
fixes #104

### CODE REVIEW

File: app/gosqs/message_attributes.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+func extractMessageAttributes(req *http.Request) map[string]MessageAttributeValue {
+	attributes := make(map[string]MessageAttributeValue)
+
+	for i := 1; true; i++ {		
+		name := req.FormValue(fmt.Sprintf("MessageAttribute.%d.Name", i))
+		if name == "" {
+			break
+		}
+		
+		dataType := req.FormValue(fmt.Sprintf("MessageAttribute.%d.Value.DataType", i))
```
Comment (p4tin) [inline app/gosqs/message_attributes.go]: I think we should check for the dataType not being present.

File: app/gosqs/message_attributes.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+		// StringListValue and BinaryListValue is currently not implemented
+		for _, valueKey := range [...]string{"StringValue", "BinaryValue"} {
```
Comment (p4tin) [inline app/gosqs/message_attributes.go]: Same comment for the Value not being present.

File: app/gosqs/message_attributes.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		}
+	}
+
+	return attributes
+}
+
+func hashAttributes(attributes map[string]MessageAttributeValue) string {
+	hasher := md5.New()
+
+	keys := sortedKeys(attributes)
```
Comment (p4tin) [inline app/gosqs/message_attributes.go]: Is there a reason for sorting the keys?

Comment (ulich) [inline app/gosqs/message_attributes.go]: iterating over the keys of a map is not deterministic. the md5 checksum would be different. I took this from the java sdk: https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-sqs/src/main/java/com/amazonaws/services/sqs/MessageMD5ChecksumHandler.java#L215

**Comment by ulich** [comment]
I did the requested changes. I added a log message, when DataType or the value does not exist. Cannot do anything properly in that case. The MD5 checksum will probably be wrong in such case.

**Comment by p4tin** [comment]
Ya I guess in such a case it's really user error but I wanted at least to know about it in the logs so they can figure out what they did wrong :)

**Comment by p4tin** [comment]
I will create a docker image later today

**Pull request merged**